sympy/core/function.py

```diff
@@ -1102,8 +1102,17 @@ def __new__(cls, expr, *variables, **assumptions):
         from sympy.matrices.common import MatrixCommon
         from sympy import Integer
         from sympy.tensor.array import Array, NDimArray, derive_by_array
+        from sympy.utilities.misc import filldedent
 
         expr = sympify(expr)
+        try:
+            has_symbol_set = isinstance(expr.free_symbols, set)
+        except AttributeError:
+            has_symbol_set = False
+        if not has_symbol_set:
+            raise ValueError(filldedent('''
+                Since there are no variables in the expression %s,
+                it cannot be differentiated.''' % expr))
```

Comment:
There is a performance regression from this pull request, `large_exprs.TimeLargeExpressionOperations.time_jacobian_wrt_symbols`, which I'm guessing (haven't verified yet) is due to this line. In general, you should avoid putting the string form of an expression in an error message, because then it slows things down for very large expressions (like the one from that benchmark), and the exception might not ever even see its way to the user if it is caught somewhere else internally.